components/sup/src/manager/file_watcher.rs

```diff
@@ -1,3464 +1,1244 @@
-            }
-            PathsAction::SettlePath(ref p) => {
-                format!("SettlePath({})", its!(p, spaces, repeat + 1))
-            }
-            PathsAction::ProcessPathAfterSettle(ref a) => {
-                format!("ProcessPathAfterSettle({})", its!(a, spaces, repeat + 1))
+    /// The run loop for FileWatcher instances.
+    pub fn run(&mut self) -> Result<()> {
+        let loop_value: liveliness_checker::ThreadUnregistered<_, _> = loop {
+            let checked_thread = liveliness_checker::mark_thread_alive();
+            if let result @ Err(_) = self.single_iteration() {
+                break checked_thread.unregister(result);
             }
-            PathsAction::RestartWatching => "RestartWatching".to_string(),
-        }
+            thread::sleep(Duration::from_secs(1));
+        };
+        loop_value.into_result()
     }
-}
-
-// Both branch result and leaf result are about the status of adding
-// new path to be watched. Branch is about symlinks and directories,
-// leaves - about regular files, missing regular files and missing
-// directories.
-//
-// TODO(asymmetric): This could be renamed to `BranchStatus`.
-enum BranchResult {
-    // The path already existed - may happen when dealing with
-    // symlinks in the path.
-    AlreadyExists,
-    // New path in a known directory.
-    NewInOldDirectory(ChainLinkInfo),
-    // New path in an unknown directory.
-    NewInNewDirectory(ChainLinkInfo, PathBuf),
-}
 
-impl IndentedToString for BranchResult {
-    fn indented_to_string(&self, spaces: &str, repeat: usize) -> String {
-        match self {
-            BranchResult::AlreadyExists => "AlreadyExists".to_string(),
-            BranchResult::NewInOldDirectory(ref i) => {
-                format!("NewInOldDirectory({})", its!(i, spaces, repeat))
-            }
-            BranchResult::NewInNewDirectory(ref i, ref p) => {
-                format!("NewInNewDirectory({}, {})",
-                        its!(i, spaces, repeat),
-                        p.to_string_lossy())
+    /// Called in our run loop, this is a single iteration wherein the FileWatcher captures its
+    /// latest observation of the WatchedFile and issues callback based on previous vs. current
+    /// state and/or notiications from the notify crate.
+    pub fn single_iteration(&mut self) -> Result<()> {
+        if self.send_initial_event {
+            if self.watched_file.exists {
+                self.callbacks.file_appeared(&self.watched_file.path);
             }
+            self.send_initial_event = false;
+            return Ok(());
         }
-    }
-}
 
-// See `BranchResult`.
-enum LeafResult {
-    // New path in a known directory.
-    NewInOldDirectory(ChainLinkInfo),
-    // New path in an unknown directory.
-    NewInNewDirectory(ChainLinkInfo, PathBuf),
-}
+        let prev_existence = self.watched_file.exists;
+        let next_existence = self.watched_file.assess_existence();
```

Comment:
maybe change to `self.watched_file.path.is_file()` and drop the `assess_existence` function.

Reply:
While we were chatting on a video call we decided it was OK to keep this.